spring-cloud-gcp-pubsub/src/main/java/com/google/cloud/spring/pubsub/core/PubSubConfiguration.java

```diff
@@ -20,18 +20,33 @@
 import com.google.api.gax.rpc.StatusCode.Code;
 import com.google.cloud.spring.pubsub.support.PubSubSubscriptionUtils;
 import com.google.pubsub.v1.ProjectSubscriptionName;
-import java.util.concurrent.ConcurrentHashMap;
-import java.util.concurrent.ConcurrentMap;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.util.Assert;
 
 /** Properties for Publisher or Subscriber specific configurations. */
 public class PubSubConfiguration {
 
+  private static final Logger logger = LoggerFactory.getLogger(PubSubConfiguration.class);
+
   /** Default number of executor threads. */
   public static final int DEFAULT_EXECUTOR_THREADS = 4;
 
   private static final Long DEFAULT_MAX_ACK_EXTENSION_PERIOD = 0L;
 
-  private final ConcurrentHashMap<String, Subscriber> subscription = new ConcurrentHashMap<>();
+  /** Automatically extracted user-provided properties. Contains only short subscription keys
+   *  user-provided properties, therefore do not use except in initialize().
+   */
+  private Map<String, Subscriber> subscription = new HashMap<>();
+
+  /** Properties keyed by fully qualified subscription name.
+   * Initialized once; effectively a singleton.
+   */
+  private Map<ProjectSubscriptionName, Subscriber> fullyQualifiedSubscriptionProperties;
```

Comment:
Maybe we can consider using an `ImmutableMap` here?

Reply:
The implementation is of an immutable map -- it's instantiated in `initialize()`, and never assigned anywhere else. What kind of immutable map is more of an implementation detail -- it could have been a Guava ImmutableMap. In this case, I went with built-in Java `UnmodifiableMap`.

I added a mention that the map is immutable to the `getFullyQualifiedSubscriberProperties()` javadoc.

Reply:
Thanks for the explanation. That makes sense! Using a UnmodifiableMap is also a valid approach.